// @lyket/react/dist/react.esm.js
import React, { useState, useEffect, useRef, useCallback, useContext } from 'react';
import { jsx, ThemeProvider } from 'theme-ui';
import { keyframes } from '@emotion/core';
import { camelizeKeys } from 'humps';

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

function _inheritsLoose(subClass, superClass) {
  subClass.prototype = Object.create(superClass.prototype);
  subClass.prototype.constructor = subClass;
  subClass.__proto__ = superClass;
}

function _getPrototypeOf(o) {
  _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) {
    return o.__proto__ || Object.getPrototypeOf(o);
  };
  return _getPrototypeOf(o);
}

function _setPrototypeOf(o, p) {
  _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) {
    o.__proto__ = p;
    return o;
  };

  return _setPrototypeOf(o, p);
}

function _isNativeReflectConstruct() {
  if (typeof Reflect === "undefined" || !Reflect.construct) return false;
  if (Reflect.construct.sham) return false;
  if (typeof Proxy === "function") return true;

  try {
    Date.prototype.toString.call(Reflect.construct(Date, [], function () {}));
    return true;
  } catch (e) {
    return false;
  }
}

function _construct(Parent, args, Class) {
  if (_isNativeReflectConstruct()) {
    _construct = Reflect.construct;
  } else {
    _construct = function _construct(Parent, args, Class) {
      var a = [null];
      a.push.apply(a, args);
      var Constructor = Function.bind.apply(Parent, a);
      var instance = new Constructor();
      if (Class) _setPrototypeOf(instance, Class.prototype);
      return instance;
    };
  }

  return _construct.apply(null, arguments);
}

function _isNativeFunction(fn) {
  return Function.toString.call(fn).indexOf("[native code]") !== -1;
}

function _wrapNativeSuper(Class) {
  var _cache = typeof Map === "function" ? new Map() : undefined;

  _wrapNativeSuper = function _wrapNativeSuper(Class) {
    if (Class === null || !_isNativeFunction(Class)) return Class;

    if (typeof Class !== "function") {
      throw new TypeError("Super expression must either be null or a function");
    }

    if (typeof _cache !== "undefined") {
      if (_cache.has(Class)) return _cache.get(Class);

      _cache.set(Class, Wrapper);
    }

    function Wrapper() {
      return _construct(Class, arguments, _getPrototypeOf(this).constructor);
    }

    Wrapper.prototype = Object.create(Class.prototype, {
      constructor: {
        value: Wrapper,
        enumerable: false,
        writable: true,
        configurable: true
      }
    });
    return _setPrototypeOf(Wrapper, Class);
  };

  return _wrapNativeSuper(Class);
}

var getSessionId = function getSessionId() {
  if (typeof window === 'undefined') {
    throw new Error('Session ID cannot be generated server-side!');
  } //TODO: remove sessionId legacy name


  var legacySessionId = window.localStorage.getItem('sessionId');

  if (legacySessionId) {
    window.localStorage.setItem('lyket-session-id', legacySessionId);
    window.localStorage.removeItem('sessionId');
  }

  var sessionId = window.localStorage.getItem('lyket-session-id');

  if (sessionId) {
    return sessionId;
  }

  var newSessionId = Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15);
  window.localStorage.setItem('lyket-session-id', newSessionId);
  return newSessionId;
};

// A type of promise-like that resolves synchronously and supports only one observer
var _iteratorSymbol = /*#__PURE__*/typeof Symbol !== "undefined" ? Symbol.iterator || (Symbol.iterator = /*#__PURE__*/Symbol("Symbol.iterator")) : "@@iterator"; // Asynchronously iterate through an object's values
var _asyncIteratorSymbol = /*#__PURE__*/typeof Symbol !== "undefined" ? Symbol.asyncIterator || (Symbol.asyncIterator = /*#__PURE__*/Symbol("Symbol.asyncIterator")) : "@@asyncIterator"; // Asynchronously iterate on a value using it's async iterator if present, or its synchronous iterator if missing

function _catch(body, recover) {
  try {
    var result = body();
  } catch (e) {
    return recover(e);
  }

  if (result && result.then) {
    return result.then(void 0, recover);
  }

  return result;
} // Asynchronously await a promise and pass the result to a finally continuation

function lazyLoadScript(src, opts) {
  if (opts === void 0) {
    opts = {};
  }

  return new Promise(function (resolve, reject) {
    var _opts = opts,
        async = _opts.async,
        integrity = _opts.integrity,
        type = _opts.type;
    var sc = document.querySelector("script[src=\"" + src + "\"]");

    if (sc) {
      resolve(sc);
      return;
    }

    var script = document.createElement('script');
    script.src = src;
    if (async) script.setAttribute('async', 'true');
    if (integrity) script.setAttribute('integrity', integrity);
    if (type) script.setAttribute('type', type);

    script.onload = function () {
      return resolve(script);
    };

    script.onerror = function (event) {
      return reject(event);
    };

    document.body.appendChild(script);
  });
}

function generateRecaptchaToken(siteKey, action) {
  return lazyLoadScript("https://www.google.com/recaptcha/api.js?render=" + siteKey).then(function () {
    return new Promise(function (resolve, reject) {
      grecaptcha.ready(function () {
        try {
          var _temp2 = _catch(function () {
            return Promise.resolve(grecaptcha.execute(siteKey, {
              action: action
            })).then(function (token) {
              resolve(token);
            });
          }, function (e) {
            console.error('recaptcha error!', e);
            reject(e);
          });

          return Promise.resolve(_temp2 && _temp2.then ? _temp2.then(function () {}) : void 0);
        } catch (e) {
          return Promise.reject(e);
        }
      });
    });
  });
}

var ApiError = /*#__PURE__*/function (_Error) {
  _inheritsLoose(ApiError, _Error);

  function ApiError(requestInfo, requestInit, status, errors) {
    var _this;

    _this = _Error.call(this, "Failed API request to " + requestInfo + " with status " + status) || this;
    _this.requestInfo = requestInfo;
    _this.requestInit = requestInit;
    _this.status = status;
    _this.errors = errors;
    return _this;
  }

  return ApiError;
}( /*#__PURE__*/_wrapNativeSuper(Error));

var now = function now() {
  return new Date().getTime();
};

function debounceCollect(func, wait) {
  var call = function call() {
    try {
      return Promise.resolve(func.call(null, args)).then(function (results) {
        for (var i = 0; i < results.length; i++) {
          deferreds[i].resolve(results[i]);
        }
      });
    } catch (e) {
      return Promise.reject(e);
    }
  };

  var onTimeout = function onTimeout() {
    try {
      var elapsed = now() - timestamp;

      var _temp2 = function () {
        if (elapsed < wait && elapsed > 0) {
          timer = setTimeout(onTimeout, wait - elapsed);
        } else {
          timer = null;
          return Promise.resolve(call()).then(function () {
            if (!timer) reset();
          });
        }
      }();

      return Promise.resolve(_temp2 && _temp2.then ? _temp2.then(function () {}) : void 0);
    } catch (e) {
      return Promise.reject(e);
    }
  };

  var timer;
  var timestamp;
  var args = [];
  var deferreds = [];

  function reset() {
    args = [];
    deferreds = [];
  }

  return function debounced() {
    for (var _len = arguments.length, callArgs = new Array(_len), _key = 0; _key < _len; _key++) {
      callArgs[_key] = arguments[_key];
    }

    args.push(callArgs);
    timestamp = now();
    if (!timer) timer = setTimeout(onTimeout, wait);
    return new Promise(function (resolve, reject) {
      deferreds.push({
        resolve: resolve,
        reject: reject
      });
    });
  };
}

var LikeButtonClient = /*#__PURE__*/function () {
  function LikeButtonClient(client) {
    this.client = client;
  }

  var _proto = LikeButtonClient.prototype;

  _proto.press = function press(_ref) {
    var id = _ref.id,
        namespace = _ref.namespace;
    return this.client.put(namespace ? "/like-buttons/" + namespace + "/" + id + "/press" : "/like-buttons/" + id + "/press", 'press');
  };

  _proto.info = function info(_ref2) {
    var id = _ref2.id,
        namespace = _ref2.namespace;
    return this.client.enqueueToBatch(namespace ? "/like-buttons/" + namespace + "/" + id : "/like-buttons/" + id);
  };

  return LikeButtonClient;
}();

var ClapButtonClient = /*#__PURE__*/function () {
  function ClapButtonClient(client) {
    this.client = client;
  }

  var _proto2 = ClapButtonClient.prototype;

  _proto2.press = function press(_ref3) {
    var id = _ref3.id,
        namespace = _ref3.namespace;
    return this.client.put(namespace ? "/clap-buttons/" + namespace + "/" + id + "/press" : "/clap-buttons/" + id + "/press", 'press');
  };

  _proto2.info = function info(_ref4) {
    var id = _ref4.id,
        namespace = _ref4.namespace;
    return this.client.enqueueToBatch(namespace ? "/clap-buttons/" + namespace + "/" + id : "/clap-buttons/" + id);
  };

  return ClapButtonClient;
}();

var UpdownButtonClient = /*#__PURE__*/function () {
  function UpdownButtonClient(client) {
    this.client = client;
  }

  var _proto3 = UpdownButtonClient.prototype;

  _proto3.info = function info(_ref5) {
    var id = _ref5.id,
        namespace = _ref5.namespace;
    return this.client.enqueueToBatch(namespace ? "/updown-buttons/" + namespace + "/" + id : "/updown-buttons/" + id);
  };

  _proto3.pressUp = function pressUp(_ref6) {
    var id = _ref6.id,
        namespace = _ref6.namespace;
    return this.client.put(namespace ? "/updown-buttons/" + namespace + "/" + id + "/press-up" : "/updown-buttons/" + id + "/press-up", 'pressUp');
  };

  _proto3.pressDown = function pressDown(_ref7) {
    var id = _ref7.id,
        namespace = _ref7.namespace;
    return this.client.put(namespace ? "/updown-buttons/" + namespace + "/" + id + "/press-down" : "/updown-buttons/" + id + "/press-down", 'pressDown');
  };

  return UpdownButtonClient;
}();

var Client = /*#__PURE__*/function () {
  function Client(_ref8) {
    var apiKey = _ref8.apiKey,
        recaptchaSiteKey = _ref8.recaptchaSiteKey,
        baseUrl = _ref8.baseUrl,
        disableSessionId = _ref8.disableSessionId;
    this.apiKey = apiKey;
    this.recaptchaSiteKey = recaptchaSiteKey;
    this.disableSessionId = disableSessionId;
    this.baseUrl = baseUrl || 'https://api.lyket.dev';
    this.likeButtons = new LikeButtonClient(this);
    this.clapButtons = new ClapButtonClient(this);
    this.updownButtons = new UpdownButtonClient(this);
    this.enqueueToBatch = debounceCollect(this.batch.bind(this), 500);
  }

  var _proto4 = Client.prototype;

  _proto4.get = function get(url) {
    return this.request(url);
  };

  _proto4.batch = function batch(calls) {
    try {
      var _this2 = this;

      // return Promise.all(calls.map(args => this.request(args[0])));
      var url = '/buttons/batch';
      return Promise.resolve(_this2.request(url, {
        method: 'POST',
        body: {
          type: 'batch',
          data: {
            attributes: {
              urls: calls.map(function (args) {
                return args[0];
              })
            }
          }
        }
      })).then(function (result) {
        return result.data.attributes.responses;
      });
    } catch (e) {
      return Promise.reject(e);
    }
  };

  _proto4.put = function put(url, recaptchaAction) {
    return this.request(url, {
      method: 'PUT',
      recaptchaAction: recaptchaAction
    });
  };

  _proto4.request = function request(input, init) {
    try {
      var _temp5 = function _temp5() {
        var url = _this4.baseUrl + "/v1" + input;
        var requestInit = init ? _extends({}, init, {
          headers: _extends({}, defaultHeaders, init.headers),
          body: init.body && JSON.stringify(init.body)
        }) : {
          headers: defaultHeaders
        };
        return Promise.resolve(fetch(url, requestInit)).then(function (response) {
          var _exit = false;

          function _temp2(_result) {
            return _exit ? _result : response.json();
          }

          var _temp = function () {
            if (response.status < 200 || response.status >= 300) {
              return Promise.resolve(response.json()).then(function (errorMessages) {
                throw new ApiError(url, requestInit, response.status, errorMessages.errors);
              });
            }
          }();

          return _temp && _temp.then ? _temp.then(_temp2) : _temp2(_temp);
        });
      };

      var _this4 = this;

      if (typeof window === 'undefined') {
        throw new Error('Client is supposed to be used client-side only!');
      }

      var defaultHeaders = {
        Accept: 'application/json',
        'Content-Type': 'application/json',
        Authorization: "Bearer " + _this4.apiKey
      };

      if (!_this4.disableSessionId) {
        defaultHeaders['x-session-id'] = getSessionId();
      }

      var _temp6 = function () {
        if (_this4.recaptchaSiteKey && init && init.recaptchaAction) {
          return Promise.resolve(generateRecaptchaToken(_this4.recaptchaSiteKey, init.recaptchaAction)).then(function (_generateRecaptchaTok) {
            defaultHeaders['x-recaptcha-token'] = _generateRecaptchaTok;
          });
        }
      }();

      return Promise.resolve(_temp6 && _temp6.then ? _temp6.then(_temp5) : _temp5(_temp6));
    } catch (e) {
      return Promise.reject(e);
    }
  };

  return Client;
}();

var ClientContext = /*#__PURE__*/React.createContext(null);

var defaultTheme = {
  colors: {
    icon: '#292929',
    background: 'transparent',
    text: '#292929',
    primary: '#BFFCED',
    secondary: '#FBDEFB',
    highlight: '#e095ed'
  },
  fonts: {
    body: 'inherit'
  }
};

var clientInstances = {};

var getClientInstanceForSettings = function getClientInstanceForSettings(settings) {
  var key = JSON.stringify(settings);

  if (clientInstances[key]) {
    return clientInstances[key];
  }

  var client = new Client(settings);
  clientInstances[key] = client;
  return client;
};

var Provider = function Provider(_ref) {
  var apiKey = _ref.apiKey,
      recaptchaSiteKey = _ref.recaptchaSiteKey,
      disableSessionId = _ref.disableSessionId,
      baseUrl = _ref.baseUrl,
      children = _ref.children,
      customTheme = _ref.theme;

  var _useState = useState(getClientInstanceForSettings({
    apiKey: apiKey,
    recaptchaSiteKey: recaptchaSiteKey,
    disableSessionId: disableSessionId,
    baseUrl: baseUrl
  })),
      client = _useState[0],
      setClient = _useState[1];

  useEffect(function () {
    setClient(getClientInstanceForSettings({
      apiKey: apiKey,
      recaptchaSiteKey: recaptchaSiteKey,
      disableSessionId: disableSessionId,
      baseUrl: baseUrl
    }));
  }, [apiKey, disableSessionId, recaptchaSiteKey, baseUrl]);
  var theme = customTheme || {
    colors: {},
    fonts: {}
  };

  var colors = _extends({}, defaultTheme.colors, theme.colors);

  var fonts = _extends({}, defaultTheme.fonts, theme.fonts);

  return jsx(ThemeProvider, {
    theme: {
      useBodyStyles: false,
      useLocalStorage: false,
      colors: colors,
      fonts: fonts,
      styles: {}
    }
  }, jsx(ClientContext.Provider, {
    value: client
  }, children));
};

var useSafeEffect = function useSafeEffect(effect, deps) {
  var setState = useState()[1];
  return useEffect(function () {
    var safeRunner = function safeRunner() {
      try {
        var _temp2 = _catch(function () {
          return Promise.resolve(effect()).then(function () {});
        }, function (e) {
          setState(function () {
            throw e;
          });
        });

        return Promise.resolve(_temp2 && _temp2.then ? _temp2.then(function () {}) : void 0);
      } catch (e) {
        return Promise.reject(e);
      }
    };

    safeRunner(); // eslint-disable-next-line
  }, [].concat(deps || [], [setState]));
};

var ClapFull = function ClapFull(_ref) {
  var className = _ref.className;
  return React.createElement("svg", {
    viewBox: "0 0 29 29",
    "aria-label": "clap",
    className: className
  }, React.createElement("path", {
    d: "M13.74 1l.76 2.97.76-2.97zM18.63 2.22l-1.43-.47-.4 3.03zM11.79 1.75l-1.43.47 1.84 2.56zM24.47 14.3L21.45 9c-.29-.43-.69-.7-1.12-.78a1.16 1.16 0 0 0-.91.22c-.3.23-.48.52-.54.84l.05.07 2.85 5c1.95 3.56 1.32 6.97-1.85 10.14a8.46 8.46 0 0 1-.55.5 5.75 5.75 0 0 0 3.36-1.76c3.26-3.27 3.04-6.75 1.73-8.91M14.58 10.89c-.16-.83.1-1.57.7-2.15l-2.5-2.49c-.5-.5-1.38-.5-1.88 0-.18.18-.27.4-.33.63l4.01 4z"
  }), React.createElement("path", {
    d: "M17.81 10.04a1.37 1.37 0 0 0-.88-.6.81.81 0 0 0-.64.15c-.18.13-.71.55-.24 1.56l1.43 3.03a.54.54 0 1 1-.87.61L9.2 7.38a.99.99 0 1 0-1.4 1.4l4.4 4.4a.54.54 0 1 1-.76.76l-4.4-4.4L5.8 8.3a.99.99 0 0 0-1.4 0 .98.98 0 0 0 0 1.39l1.25 1.24 4.4 4.4a.54.54 0 0 1 0 .76.54.54 0 0 1-.76 0l-4.4-4.4a1 1 0 0 0-1.4 0 .98.98 0 0 0 0 1.4l1.86 1.85 2.76 2.77a.54.54 0 0 1-.76.76L4.58 15.7a.98.98 0 0 0-1.4 0 .99.99 0 0 0 0 1.4l5.33 5.32c3.37 3.37 6.64 4.98 10.49 1.12 2.74-2.74 3.27-5.54 1.62-8.56l-2.8-4.94z"
  }));
};

var style = {
  root: {
    color: 'inherit',
    fontFamily: 'inherit',
    display: 'inline-flex',
    alignItems: 'center'
  },
  counter: {
    ml: '.5em',
    color: 'text',
    fontFamily: 'body'
  },
  icon: {
    height: '1.6em',
    display: 'block',
    margin: '0 auto',
    fill: 'icon'
  },
  ring: {
    opacity: '0',
    width: "90%",
    height: "90%",
    borderRadius: "3em",
    border: "1px solid",
    borderColor: 'highlight',
    boxSizing: 'border-box',
    position: 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translate(-50%, -50%)'
  },
  button: {
    border: 'none',
    outline: 'none',
    boxShadow: '0 0 0',
    transform: 'translate3d(0,0,0)',
    cursor: 'pointer',
    appearance: 'none',
    position: 'relative',
    fontSize: 'inherit',
    borderRadius: '100%',
    transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
    width: '3em',
    height: '3em',
    padding: '0',
    lineHeight: '1em',
    bg: 'background',
    ':hover': {
      transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
      opacity: 0.75
    }
  },
  baloon: {
    opacity: '0',
    position: 'absolute',
    fontFamily: 'body',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    color: 'white',
    backgroundColor: '#292929',
    borderRadius: '100%',
    fontSize: '.7em',
    height: '2.8em',
    width: '2.8em',
    top: '-1.4em',
    left: '-1.4em'
  },
  centeredContainer: {
    position: 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translate(-50%, -50%)'
  },
  buttonContainer: {
    position: 'relative'
  }
};

var theme = {
  buttons: {
    inactive: {
      transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
      backgroundColor: 'background'
    },
    active: {
      backgroundColor: 'primary'
    }
  }
};

var baloonFade = /*#__PURE__*/keyframes({
  '0%': {
    opacity: '0',
    transform: 'translateY(-3.2em)'
  },
  '72%': {
    opacity: '1',
    transform: 'translateY(-4em)'
  },
  '100%': {
    opacity: '0',
    transform: 'translateY(-6em)'
  }
});
var iconScale = /*#__PURE__*/keyframes({
  '0%': {
    transform: 'scale(0)'
  },
  '52%': {
    transform: 'scale(1.2)'
  },
  '72%': {
    transform: 'scale(0.9)'
  },
  '100%': {
    transform: 'scale(1)'
  }
});
var ringScale = /*#__PURE__*/keyframes({
  '0%': {
    opacity: '1',
    borderWidth: '0.5em'
  },
  '52%': {
    opacity: '0',
    borderWidth: '0'
  }
});
var BALOON = {
  durationMs: 800,
  animation: baloonFade
};
var ICON = {
  durationMs: 700,
  animation: iconScale
};
var RING = {
  durationMs: 600,
  animation: ringScale
};
var Simple = function Simple(_ref) {
  var isLoading = _ref.isLoading,
      userClaps = _ref.userClaps,
      totalClaps = _ref.totalClaps,
      handlePress = _ref.handlePress,
      isCounterVisible = _ref.isCounterVisible;

  var _useState = useState(false),
      animationActive = _useState[0],
      setAnimationActive = _useState[1];

  var timeoutId = useRef(null);
  var handleClick = useCallback(function (e) {
    setAnimationActive(true);

    if (timeoutId.current) {
      clearTimeout(timeoutId.current);
    }

    timeoutId.current = setTimeout(function () {
      return setAnimationActive(false);
    }, BALOON.durationMs);
    handlePress(e);
  }, [handlePress]);

  var baloonStyle = _extends({}, style.baloon, {
    animation: animationActive ? BALOON.animation + " " + BALOON.durationMs + "ms ease forwards" : null
  });

  var iconStyle = _extends({}, style.icon, {
    animation: animationActive ? ICON.animation + " " + ICON.durationMs + "ms ease forwards" : null
  });

  var ringStyle = _extends({}, style.ring, {
    animation: animationActive ? RING.animation + " " + RING.durationMs + "ms ease forwards" : null
  });

  var buttonStyle = _extends({}, style.button, {
    variant: userClaps ? 'buttons.active' : 'buttons.inactive',
    '@media (hover: hover)': {
      ':hover': {
        bg: userClaps ? 'primary' : 'background'
      }
    }
  });

  return jsx(ThemeProvider, {
    theme: theme
  }, jsx("div", {
    sx: style.root
  }, jsx("div", {
    sx: style.buttonContainer
  }, jsx("div", {
    sx: style.centeredContainer
  }, jsx("div", {
    key: timeoutId.current && timeoutId.current.toString(),
    sx: baloonStyle
  }, userClaps)), jsx("button", {
    onClick: handleClick,
    sx: buttonStyle,
    disabled: isLoading
  }, jsx("div", {
    sx: ringStyle
  }), jsx(ClapFull, {
    sx: iconStyle
  }))), isCounterVisible && jsx("div", {
    sx: style.counter
  }, totalClaps)));
};

var ClapHeart = function ClapHeart(_ref) {
  var className = _ref.className;
  return React.createElement("svg", {
    version: "1.1",
    id: "ClapHeart",
    xmlns: "http://www.w3.org/2000/svg",
    x: "0px",
    y: "0px",
    viewBox: "0 0 110 115",
    "aria-label": "clap",
    className: className
  }, React.createElement("g", null, React.createElement("g", null, React.createElement("path", {
    d: "M42.68,27.43c-0.7,0-1.4-0.27-1.93-0.8l-5.91-5.91c-1.07-1.07-1.07-2.79,0-3.86c1.07-1.07,2.79-1.07,3.86,0 l5.91,5.91c1.07,1.07,1.07,2.79,0,3.86C44.08,27.16,43.38,27.43,42.68,27.43z"
  })), React.createElement("g", null, React.createElement("path", {
    d: "M70.43,26.52c-0.7,0-1.4-0.27-1.93-0.8c-1.07-1.07-1.07-2.79,0-3.86l5.91-5.91c1.07-1.07,2.79-1.07,3.86,0 c1.07,1.07,1.07,2.79,0,3.86l-5.91,5.91C71.82,26.25,71.13,26.52,70.43,26.52z"
  })), React.createElement("path", {
    d: "M67.23,7.48c-0.12-1.34-0.82-2.56-1.93-3.34l0,0c-1.77-1.24-4.33-0.97-5.94,0.65l-2.8,2.79l-2.79-2.79 c-1.61-1.61-4.17-1.9-5.94-0.65c-1.11,0.78-1.82,2-1.93,3.34c-0.12,1.34,0.36,2.66,1.31,3.61L53.11,17c0.86,0.86,2,1.33,3.22,1.33 c0.07,0,0.15,0,0.23-0.01c0.08,0.01,0.15,0.01,0.23,0.01c1.22,0,2.36-0.47,3.22-1.33l5.91-5.91C66.86,10.14,67.34,8.82,67.23,7.48z"
  }), React.createElement("path", {
    d: "M104.64,60.44c-0.28-1.84-1.22-3.52-2.54-4.83l-0.06-0.06c-1.02-1.02-2.25-1.75-3.6-2.15l0.16-0.16 c3.32-3.32,3.33-8.71,0-12.04l-0.21-0.21c-1-1-2.19-1.71-3.5-2.12c1.4-3.13,0.82-6.94-1.75-9.5l-0.04-0.04 c-1.71-1.71-4.03-2.72-6.45-2.65c-1.76,0.05-3.48,0.54-4.8,1.67c-0.29,0.25-2.61,2.55-3.11,3.04c-0.8,0.79-0.71,2.1,0.19,2.77 l0.88,0.66c0.61,0.46,1.47,0.4,2.01-0.14l2.29-2.29c0.77-0.77,1.79-1.16,2.8-1.16c1.01,0,2.03,0.39,2.8,1.16l0.21,0.21 c1.55,1.55,1.55,4.06,0,5.61l-11.3,11.3c-0.85-0.69-1.81-1.21-2.84-1.53c1.4-3.13,0.82-6.94-1.75-9.5l-0.21-0.21 c-1.61-1.61-3.75-2.49-6.02-2.49c-1.22,0-2.41,0.26-3.49,0.74c-0.39-1.28-1.1-2.49-2.11-3.5l-0.21-0.21 c-1.61-1.61-3.75-2.49-6.02-2.49c-2.27,0-4.41,0.89-6.02,2.49L34.35,48.43c-0.4-1.34-1.13-2.58-2.15-3.6l-1.49-1.49 c-1.26-1.26-2.94-1.96-4.73-1.96c-1.79,0-3.47,0.7-4.73,1.96l-8.49,8.49l-0.58,0.58l-3.64,3.64c-3.64,3.64-5.52,9.27-5.3,15.84 c0.32,9.4,4.72,18.68,11.76,24.83c9.22,8.05,15.07,11.87,23.37,11.87c1.2,0,2.47-0.08,3.9-0.24c9.42-1.05,16.6-7.92,17.23-8.55 l0.44-0.42c0.47-0.04,0.93-0.07,1.43-0.13c9.42-1.05,16.6-7.92,17.23-8.55l23.59-22.88l0.05-0.05 C104.16,65.87,105.05,63.2,104.64,60.44z M92.38,43.04c1.01,0,2.03,0.39,2.8,1.16l0.21,0.21c1.55,1.55,1.55,4.06,0,5.6L81.75,63.66 c-0.74-0.51-1.55-0.92-2.4-1.17l0.16-0.16c2.62-2.62,3.16-6.53,1.65-9.7l8.43-8.43C90.36,43.43,91.37,43.04,92.38,43.04z M56.33,96.56c0,0-6.37,6.37-14.55,7.28c-1.2,0.13-2.32,0.21-3.4,0.21c-6.26,0-10.96-2.53-20.38-10.75 c-11.05-9.64-13.16-27.1-6.24-34.03l3.64-3.64l0.58-0.58l8.49-8.49c0.42-0.42,0.97-0.63,1.51-0.63c0.55,0,1.1,0.21,1.51,0.63 l1.49,1.49c1.55,1.55,1.55,4.06,0,5.6l-6.13,6.2c-0.89,0.9-0.89,2.35,0.01,3.25c0.45,0.45,1.04,0.67,1.63,0.67 c0.59,0,1.18-0.22,1.63-0.67l8.38-8.38l18.7-18.7c0.77-0.77,1.79-1.16,2.8-1.16c1.01,0,2.03,0.39,2.8,1.16l0.21,0.21 c1.55,1.55,1.55,4.06,0,5.61l-17.2,17.2c-0.79,0.79-0.78,2.07,0.02,2.85c0.39,0.38,0.89,0.57,1.4,0.57c0.51,0,1.02-0.2,1.41-0.59 l20.38-20.38c0.77-0.77,1.79-1.16,2.8-1.16c1.01,0,2.03,0.39,2.8,1.16l0.21,0.21c1.55,1.55,1.55,4.06,0,5.61L50.56,67.56 c-0.79,0.79-0.78,2.07,0.02,2.85c0.39,0.38,0.89,0.57,1.4,0.57c0.51,0,1.02-0.2,1.41-0.59L70.48,53.3c0.77-0.77,1.79-1.16,2.8-1.16 c1.01,0,2.03,0.39,2.8,1.16l0.21,0.21c0.38,0.38,0.66,0.81,0.85,1.28l0.27,1.59l0.04-0.04c-0.01,1.01-0.39,2.01-1.16,2.78 L59.54,75.87c-0.91,0.91-0.91,2.4,0.02,3.3c0.45,0.44,1.04,0.66,1.62,0.66c0.59,0,1.19-0.23,1.64-0.68l11.3-11.3 c0.77-0.77,1.79-1.16,2.8-1.16c1.01,0,2.03,0.39,2.8,1.16l0.21,0.21c1.55,1.55,1.55,4.06,0,5.6L56.33,96.56z M99.03,64.57 l-23.6,22.89c0,0-3.81,3.8-9.32,5.94L83.1,76.93l0.02-0.02l0.02-0.02c1.61-1.61,2.49-3.74,2.49-6.02c0-1.3-0.3-2.55-0.84-3.68 l8.42-8.42c0.77-0.77,1.79-1.16,2.8-1.16s2.03,0.39,2.8,1.16l0.21,0.21C100.58,60.51,100.58,63.02,99.03,64.57z"
  })));
};

var style$1 = {
  root: {
    color: 'inherit',
    display: 'inline-flex',
    alignItems: 'center'
  },
  counter: {
    ml: '.4em',
    fontFamily: 'body',
    color: 'text'
  },
  icon: {
    height: '2.2em',
    transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms'
  },
  button: {
    border: 'none',
    outline: 'none',
    cursor: 'pointer',
    background: 'transparent',
    appearance: 'none',
    position: 'relative',
    fontSize: 'inherit',
    p: 0,
    '@media (hover: hover)': {
      ':hover': {
        svg: {
          transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
          transform: 'scale(1.05)'
        }
      },
      ':active': {
        svg: {
          transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
          transform: 'scale(1)'
        }
      }
    }
  },
  baloon: {
    opacity: '0',
    position: 'absolute',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    color: 'white',
    backgroundColor: '#292929',
    fontFamily: 'body',
    borderRadius: '100%',
    fontSize: '.6em',
    height: '2.4em',
    width: '2.4em',
    top: '-1.2em',
    left: '-1.2em'
  },
  centeredContainer: {
    position: 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translate(-50%, -50%)'
  },
  buttonContainer: {
    position: 'relative'
  }
};

var baloonFade$1 = /*#__PURE__*/keyframes({
  '0%': {
    opacity: '0.2',
    transform: 'translateY(-3.2em)'
  },
  '72%': {
    opacity: '1',
    transform: 'translateY(-3.8em)'
  },
  '100%': {
    opacity: '0',
    transform: 'translateY(-5.2em)'
  }
});
var triangleFade = /*#__PURE__*/keyframes({
  '0%': {
    opacity: '0',
    transform: 'translateY(-0.5em) scale(1)'
  },
  '32%': {
    opacity: '1',
    transform: 'translateY(-1em) scale(1.3)'
  },
  '100%': {
    opacity: '0',
    transform: 'translateY(-2em) scale(1.6)'
  }
});
var iconScale$1 = /*#__PURE__*/keyframes({
  '0%': {
    transform: 'scale(1)'
  },
  '32%': {
    transform: 'scale(1.1)'
  },
  '100%': {
    transform: 'scale(1.05)'
  }
});
var BALOON$1 = {
  durationMs: 800,
  animation: baloonFade$1
};
var TRIANGLES = {
  count: 5,
  size: 0.2,
  color: 'secondary',
  durationMs: 300,
  animation: triangleFade
};
var CIRCLES = {
  count: 5,
  size: 0.05,
  color: 'primary',
  durationMs: 500,
  animation: triangleFade
};
var ICON$1 = {
  durationMs: CIRCLES.durationMs,
  animation: iconScale$1
};
var Heart = function Heart(_ref) {
  var isLoading = _ref.isLoading,
      userClaps = _ref.userClaps,
      totalClaps = _ref.totalClaps,
      handlePress = _ref.handlePress,
      isCounterVisible = _ref.isCounterVisible;

  var _useState = useState(false),
      animationActive = _useState[0],
      setAnimationActive = _useState[1];

  var triangles = [];

  for (var i = 0; i < TRIANGLES.count; i++) {
    var trianglesStyle = {
      width: '0',
      height: '0',
      opacity: '0',
      borderLeft: TRIANGLES.size + "em solid transparent",
      borderRight: TRIANGLES.size + "em solid transparent",
      borderTop: TRIANGLES.size + "em solid " + TRIANGLES.color,
      position: 'absolute',
      animation: animationActive ? triangleFade + " " + TRIANGLES.durationMs + "ms ease forwards" : null
    };
    var trianglesContainerStyle = {
      transform: "rotate(" + 360 / TRIANGLES.count * i + "deg)",
      transformOrigin: TRIANGLES.size + "em 0",
      marginLeft: "-" + TRIANGLES.size + "em"
    };
    triangles.push(jsx("div", {
      key: i,
      sx: trianglesContainerStyle
    }, jsx("div", {
      sx: trianglesStyle
    })));
  }

  var circles = [];

  for (var _i = 0; _i < CIRCLES.count; _i++) {
    var circlesStyle = {
      width: '0',
      height: '0',
      opacity: '0',
      padding: CIRCLES.size + "em",
      backgroundColor: "" + CIRCLES.color,
      borderRadius: '100%',
      position: 'absolute',
      animation: animationActive ? triangleFade + " " + CIRCLES.durationMs + "ms ease forwards" : null
    };
    var circleContainerStyle = {
      transform: "rotate(" + 360 / CIRCLES.count * _i + "deg)",
      transformOrigin: CIRCLES.size + "em 0",
      marginLeft: "-" + CIRCLES.size + "em"
    };
    circles.push(jsx("div", {
      key: _i,
      sx: circleContainerStyle
    }, jsx("div", {
      sx: circlesStyle
    })));
  }

  var timeoutId = useRef(null);
  var handleClick = useCallback(function (e) {
    setAnimationActive(true);

    if (timeoutId.current) {
      clearTimeout(timeoutId.current);
    }

    timeoutId.current = setTimeout(function () {
      return setAnimationActive(false);
    }, BALOON$1.durationMs);
    handlePress(e);
  }, [handlePress]);

  var iconStyle = _extends({}, style$1.icon, {
    fill: userClaps && userClaps > 0 ? 'highlight' : 'icon',
    animation: animationActive ? ICON$1.animation + " " + ICON$1.durationMs + "ms ease forwards" : null
  });

  var baloonStyle = _extends({}, style$1.baloon, {
    animation: animationActive ? BALOON$1.animation + " " + BALOON$1.durationMs + "ms ease forwards" : null
  });

  return jsx("div", {
    sx: style$1.root
  }, jsx("div", {
    sx: style$1.buttonContainer
  }, jsx("div", {
    sx: style$1.centeredContainer
  }, jsx("div", {
    key: timeoutId.current && timeoutId.current.toString(),
    sx: baloonStyle
  }, userClaps)), jsx("button", {
    onClick: handleClick,
    sx: style$1.button,
    disabled: isLoading
  }, jsx("div", {
    sx: style$1.centeredContainer
  }, " ", triangles, " "), jsx("div", {
    sx: style$1.centeredContainer
  }, " ", circles, " "), jsx(ClapHeart, {
    sx: iconStyle
  }))), isCounterVisible && jsx("div", {
    sx: style$1.counter
  }, totalClaps));
};

var ClapEmpty = function ClapEmpty(_ref) {
  var className = _ref.className;
  return React.createElement("svg", {
    viewBox: "0 0 33 33",
    "aria-label": "clap",
    className: className
  }, React.createElement("path", {
    d: "M28.86 17.34l-3.64-6.4c-.3-.43-.71-.73-1.16-.8a1.12 1.12 0 0 0-.9.21c-.62.5-.73 1.18-.32 2.06l1.22 2.6 1.4 2.45c2.23 4.09 1.51 8-2.15 11.66a9.6 9.6 0 0 1-.8.71 6.53 6.53 0 0 0 4.3-2.1c3.82-3.82 3.57-7.87 2.05-10.39zm-6.25 11.08c3.35-3.35 4-6.78 1.98-10.47L21.2 12c-.3-.43-.71-.72-1.16-.8a1.12 1.12 0 0 0-.9.22c-.62.49-.74 1.18-.32 2.06l1.72 3.63a.5.5 0 0 1-.81.57l-8.91-8.9a1.33 1.33 0 0 0-1.89 1.88l5.3 5.3a.5.5 0 0 1-.71.7l-5.3-5.3-1.49-1.49c-.5-.5-1.38-.5-1.88 0a1.34 1.34 0 0 0 0 1.89l1.49 1.5 5.3 5.28a.5.5 0 0 1-.36.86.5.5 0 0 1-.36-.15l-5.29-5.29a1.34 1.34 0 0 0-1.88 0 1.34 1.34 0 0 0 0 1.89l2.23 2.23L9.3 21.4a.5.5 0 0 1-.36.85.5.5 0 0 1-.35-.14l-3.32-3.33a1.33 1.33 0 0 0-1.89 0 1.32 1.32 0 0 0-.39.95c0 .35.14.69.4.94l6.39 6.4c3.53 3.53 8.86 5.3 12.82 1.35zM12.73 9.26l5.68 5.68-.49-1.04c-.52-1.1-.43-2.13.22-2.89l-3.3-3.3a1.34 1.34 0 0 0-1.88 0 1.33 1.33 0 0 0-.4.94c0 .22.07.42.17.61zm14.79 19.18a7.46 7.46 0 0 1-6.41 2.31 7.92 7.92 0 0 1-3.67.9c-3.05 0-6.12-1.63-8.36-3.88l-6.4-6.4A2.31 2.31 0 0 1 2 19.72a2.33 2.33 0 0 1 1.92-2.3l-.87-.87a2.34 2.34 0 0 1 0-3.3 2.33 2.33 0 0 1 1.24-.64l-.14-.14a2.34 2.34 0 0 1 0-3.3 2.39 2.39 0 0 1 3.3 0l.14.14a2.33 2.33 0 0 1 3.95-1.24l.09.09c.09-.42.29-.83.62-1.16a2.34 2.34 0 0 1 3.3 0l3.38 3.39a2.17 2.17 0 0 1 1.27-.17c.54.08 1.03.35 1.45.76.1-.55.41-1.03.9-1.42a2.12 2.12 0 0 1 1.67-.4 2.8 2.8 0 0 1 1.85 1.25l3.65 6.43c1.7 2.83 2.03 7.37-2.2 11.6zM13.22.48l-1.92.89 2.37 2.83-.45-3.72zm8.48.88L19.78.5l-.44 3.7 2.36-2.84zM16.5 3.3L15.48 0h2.04L16.5 3.3z",
    fillRule: "evenodd"
  }));
};

var ClapFull$1 = function ClapFull(_ref) {
  var className = _ref.className;
  return React.createElement("svg", {
    viewBox: "0 0 29 29",
    "aria-label": "clap",
    className: className
  }, React.createElement("path", {
    d: "M13.74 1l.76 2.97.76-2.97zM18.63 2.22l-1.43-.47-.4 3.03zM11.79 1.75l-1.43.47 1.84 2.56zM24.47 14.3L21.45 9c-.29-.43-.69-.7-1.12-.78a1.16 1.16 0 0 0-.91.22c-.3.23-.48.52-.54.84l.05.07 2.85 5c1.95 3.56 1.32 6.97-1.85 10.14a8.46 8.46 0 0 1-.55.5 5.75 5.75 0 0 0 3.36-1.76c3.26-3.27 3.04-6.75 1.73-8.91M14.58 10.89c-.16-.83.1-1.57.7-2.15l-2.5-2.49c-.5-.5-1.38-.5-1.88 0-.18.18-.27.4-.33.63l4.01 4z"
  }), React.createElement("path", {
    d: "M17.81 10.04a1.37 1.37 0 0 0-.88-.6.81.81 0 0 0-.64.15c-.18.13-.71.55-.24 1.56l1.43 3.03a.54.54 0 1 1-.87.61L9.2 7.38a.99.99 0 1 0-1.4 1.4l4.4 4.4a.54.54 0 1 1-.76.76l-4.4-4.4L5.8 8.3a.99.99 0 0 0-1.4 0 .98.98 0 0 0 0 1.39l1.25 1.24 4.4 4.4a.54.54 0 0 1 0 .76.54.54 0 0 1-.76 0l-4.4-4.4a1 1 0 0 0-1.4 0 .98.98 0 0 0 0 1.4l1.86 1.85 2.76 2.77a.54.54 0 0 1-.76.76L4.58 15.7a.98.98 0 0 0-1.4 0 .99.99 0 0 0 0 1.4l5.33 5.32c3.37 3.37 6.64 4.98 10.49 1.12 2.74-2.74 3.27-5.54 1.62-8.56l-2.8-4.94z"
  }));
};

var style$2 = {
  root: {
    color: 'inherit',
    fontFamily: 'inherit',
    display: 'inline-flex',
    alignItems: 'center'
  },
  counter: {
    ml: '.6em',
    fontFamily: 'body'
  },
  icon: {
    height: '2em',
    fill: 'icon',
    transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms'
  },
  button: {
    border: 'none',
    outline: 'none',
    cursor: 'pointer',
    background: 'transparent',
    appearance: 'none',
    position: 'relative',
    fontSize: 'inherit',
    p: 0,
    '@media (hover: hover)': {
      ':hover': {
        transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
        opacity: 0.65
      },
      ':active': {
        svg: {
          transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
          transform: 'scale(0.9)'
        }
      }
    }
  },
  baloon: {
    opacity: '0',
    position: 'absolute',
    fontFamily: 'inherit',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    color: 'white',
    backgroundColor: '#292929',
    borderRadius: '100%',
    fontSize: '.7em',
    height: '2.4em',
    width: '2.4em',
    top: '-1.2em',
    left: '-1.2em',
    '@media (hover: hover)': {
      ':hover': {
        opacity: '1'
      }
    }
  },
  centeredContainer: {
    position: 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translate(-50%, -50%)'
  },
  buttonContainer: {
    position: 'relative'
  }
};

var baloonFade$2 = /*#__PURE__*/keyframes({
  '0%': {
    opacity: '0.5',
    transform: 'translateY(-2.6em)'
  },
  '72%': {
    opacity: '1',
    transform: 'translateY(-3.4em)'
  },
  '100%': {
    opacity: '0',
    transform: 'translateY(-5em)'
  }
});
var triangleFade$1 = /*#__PURE__*/keyframes({
  '0%': {
    opacity: '0',
    transform: 'translateY(-0.5em) scale(1)'
  },
  '32%': {
    opacity: '1',
    transform: 'translateY(-1em) scale(1.3)'
  },
  '100%': {
    opacity: '0',
    transform: 'translateY(-2em) scale(1.6)'
  }
});
var iconScale$2 = /*#__PURE__*/keyframes({
  '0%': {
    transform: 'scale(1)'
  },
  '32%': {
    transform: 'scale(1.1)'
  },
  '100%': {
    transform: 'scale(1)'
  }
});
var BALOON$2 = {
  durationMs: 800,
  animation: baloonFade$2
};
var TRIANGLES$1 = {
  count: 5,
  size: 0.2,
  color: '#ff365e',
  durationMs: 300,
  animation: triangleFade$1
};
var CIRCLES$1 = {
  count: 5,
  size: 0.05,
  color: '#4086ff',
  durationMs: 500,
  animation: triangleFade$1
};
var ICON$2 = {
  durationMs: CIRCLES$1.durationMs,
  animation: iconScale$2
};
var Medium = function Medium(_ref) {
  var isLoading = _ref.isLoading,
      userClaps = _ref.userClaps,
      totalClaps = _ref.totalClaps,
      handlePress = _ref.handlePress,
      isCounterVisible = _ref.isCounterVisible;

  var _useState = useState(false),
      animationActive = _useState[0],
      setAnimationActive = _useState[1];

  var triangles = [];

  for (var i = 0; i < TRIANGLES$1.count; i++) {
    triangles.push(jsx("div", {
      key: i,
      sx: {
        transform: "rotate(" + 360 / TRIANGLES$1.count * i + "deg)",
        transformOrigin: TRIANGLES$1.size + "em 0",
        marginLeft: "-" + TRIANGLES$1.size + "em"
      }
    }, jsx("div", {
      sx: {
        width: '0',
        height: '0',
        opacity: '0',
        borderLeft: TRIANGLES$1.size + "em solid transparent",
        borderRight: TRIANGLES$1.size + "em solid transparent",
        borderTop: TRIANGLES$1.size + "em solid " + TRIANGLES$1.color,
        position: 'absolute',
        animation: animationActive ? triangleFade$1 + " " + TRIANGLES$1.durationMs + "ms ease forwards" : null
      }
    })));
  }

  var circles = [];

  for (var _i = 0; _i < CIRCLES$1.count; _i++) {
    circles.push(jsx("div", {
      key: _i,
      sx: {
        transform: "rotate(" + 360 / CIRCLES$1.count * _i + "deg)",
        transformOrigin: CIRCLES$1.size + "em 0",
        marginLeft: "-" + CIRCLES$1.size + "em"
      }
    }, jsx("div", {
      sx: {
        width: '0',
        height: '0',
        opacity: '0',
        padding: CIRCLES$1.size + "em",
        backgroundColor: "" + CIRCLES$1.color,
        borderRadius: '100%',
        position: 'absolute',
        animation: animationActive ? triangleFade$1 + " " + CIRCLES$1.durationMs + "ms ease forwards" : null
      }
    })));
  }

  var timeoutId = useRef(null);
  var handleClick = useCallback(function (e) {
    setAnimationActive(true);

    if (timeoutId.current) {
      clearTimeout(timeoutId.current);
    }

    timeoutId.current = setTimeout(function () {
      return setAnimationActive(false);
    }, BALOON$2.durationMs);
    handlePress(e);
  }, [handlePress]);
  var Icon = userClaps && userClaps > 0 ? ClapFull$1 : ClapEmpty;

  var iconStyle = _extends({}, style$2.icon, {
    animation: animationActive ? ICON$2.animation + " " + ICON$2.durationMs + "ms ease forwards" : null
  });

  var baloonStyle = _extends({}, style$2.baloon, {
    animation: animationActive ? BALOON$2.animation + " " + BALOON$2.durationMs + "ms ease forwards" : null
  });

  return jsx("div", {
    sx: style$2.root
  }, jsx("div", {
    sx: style$2.buttonContainer
  }, jsx("div", {
    sx: style$2.centeredContainer
  }, jsx("div", {
    key: timeoutId.current && timeoutId.current.toString(),
    sx: baloonStyle
  }, userClaps)), jsx("button", {
    onClick: handleClick,
    sx: style$2.button,
    disabled: isLoading
  }, jsx("div", {
    sx: style$2.centeredContainer
  }, " ", triangles, " "), jsx("div", {
    sx: style$2.centeredContainer
  }, " ", circles, " "), jsx(Icon, {
    sx: iconStyle
  }))), isCounterVisible && jsx("div", {
    sx: style$2.counter
  }, totalClaps));
};

var templates = {
  Simple: Simple,
  Medium: Medium,
  Heart: Heart
};

var ClapButton = function ClapButton(_ref) {
  var id = _ref.id,
      namespace = _ref.namespace,
      hideCounterIfLessThan = _ref.hideCounterIfLessThan,
      children = _ref.children,
      component = _ref.component,
      onLoad = _ref.onLoad,
      onPress = _ref.onPress;
  var client = useContext(ClientContext);

  var _useState = useState(null),
      response = _useState[0],
      setResponse = _useState[1];

  useSafeEffect(function () {
    try {
      return Promise.resolve(_catch(function () {
        var _temp = function () {
          if (client) {
            return Promise.resolve(client.clapButtons.info({
              id: id,
              namespace: namespace
            })).then(function (result) {
              setResponse(result.data);

              if (onLoad) {
                onLoad(camelizeKeys(result.data));
              }
            });
          }
        }();

        if (_temp && _temp.then) return _temp.then(function () {});
      }, function (error) {
        console.error('Lyket error:', error.errors[0].message);
        throw error;
      }));
    } catch (e) {
      return Promise.reject(e);
    }
  }, [client, id, namespace, setResponse]);
  var handlePress = useCallback(function (e) {
    try {
      e.preventDefault();
      return Promise.resolve(_catch(function () {
        var _temp2 = function () {
          if (client) {
            return Promise.resolve(client.clapButtons.press({
              id: id,
              namespace: namespace
            })).then(function (result) {
              setResponse(result.data);

              if (onPress) {
                onPress(camelizeKeys(result.data));
              }
            });
          }
        }();

        if (_temp2 && _temp2.then) return _temp2.then(function () {});
      }, function (error) {
        console.error('Lyket error:', error.errors[0].message);
        throw error;
      }));
    } catch (e) {
      return Promise.reject(e);
    }
  }, [client, id, namespace, onPress]);
  var isCounterVisible = true;

  if (!response || hideCounterIfLessThan && response && response.attributes.total_claps < hideCounterIfLessThan) {
    isCounterVisible = false;
  }

  var props = {
    isLoading: !response,
    totalClaps: response && response.attributes.total_claps || 0,
    userClaps: response && response.attributes.user_claps || 0,
    handlePress: handlePress,
    isCounterVisible: isCounterVisible
  };

  if (children) {
    return children(props);
  }

  var Component = component || Simple;
  return React.createElement(Component, Object.assign({}, props));
};

ClapButton.templates = templates;

var ThumbIcon = function ThumbIcon(_ref) {
  var className = _ref.className;
  return React.createElement("svg", {
    id: "ThumUpWiki",
    viewBox: "0 0 215.34 215.34",
    className: className
  }, React.createElement("path", {
    d: "M7.67,93.896h52.077V215.34H7.67V93.896z M207.67,106.682c-7.189-7.189-14.382-14.379-21.572-21.57h-41.451l9.877-25.102  l2.801-7.119l-0.258-2.049L151,2.33L121.115,0L110.15,37.822L80.954,85.176V215.34h107.89L207.67,106.682z"
  }));
};

var style$3 = {
  root: {
    color: 'inherit',
    fontFamily: 'inherit',
    display: 'inline-flex',
    alignItems: 'center'
  },
  counter: {
    ml: '.5em',
    color: 'text',
    fontFamily: 'body'
  },
  icon: {
    height: '1em',
    display: 'block',
    margin: '0 auto',
    fill: 'icon'
  },
  ring: {
    opacity: '0',
    width: "90%",
    height: "90%",
    borderRadius: "3em",
    border: "1px solid",
    borderColor: 'highlight',
    boxSizing: 'border-box',
    position: 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translate(-50%, -50%)'
  },
  button: {
    border: 'none',
    outline: 'none',
    boxShadow: '0 0 0',
    transform: 'translate3d(0,0,0)',
    cursor: 'pointer',
    bg: 'background',
    appearance: 'none',
    position: 'relative',
    fontSize: 'inherit',
    borderRadius: '100%',
    transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
    width: '3em',
    height: '3em',
    padding: '0',
    lineHeight: '1em',
    ':hover': {
      transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
      opacity: 0.75
    }
  }
};

var theme$1 = {
  buttons: {
    inactive: {
      backgroundColor: 'background'
    },
    active: {
      transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
      backgroundColor: 'primary'
    }
  }
};

var iconScale$3 = /*#__PURE__*/keyframes({
  '0%': {
    transform: 'scale(0)'
  },
  '52%': {
    transform: 'scale(1.2)'
  },
  '72%': {
    transform: 'scale(0.9)'
  },
  '100%': {
    transform: 'scale(1)'
  }
});
var ringScale$1 = /*#__PURE__*/keyframes({
  '0%': {
    opacity: '1',
    borderWidth: '0.5em'
  },
  '52%': {
    opacity: '0',
    borderWidth: '0'
  }
});
var RING$1 = {
  durationMs: 600,
  animation: ringScale$1
};
var ICON$3 = {
  animation: iconScale$3,
  durationMs: 700
};
var Simple$1 = function Simple(_ref) {
  var isLoading = _ref.isLoading,
      userLiked = _ref.userLiked,
      totalLikes = _ref.totalLikes,
      handlePress = _ref.handlePress,
      isCounterVisible = _ref.isCounterVisible;

  var _useState = useState(false),
      animationActive = _useState[0],
      setAnimationActive = _useState[1];

  var handleClick = useCallback(function (e) {
    handlePress(e);

    if (!userLiked) {
      setAnimationActive(true);
      setTimeout(function () {
        return setAnimationActive(false);
      }, ICON$3.durationMs);
    }
  }, [userLiked, handlePress]);

  var iconStyle = _extends({}, style$3.icon, {
    animation: animationActive ? ICON$3.animation + " " + ICON$3.durationMs + "ms ease forwards" : null
  });

  var ringStyle = _extends({}, style$3.ring, {
    animation: animationActive ? RING$1.animation + " " + RING$1.durationMs + "ms ease forwards" : null
  });

  var buttonStyle = {
    variant: userLiked ? 'buttons.active' : 'buttons.inactive',
    '@media (hover: hover)': {
      ':hover': {
        bg: userLiked ? 'primary' : 'background'
      }
    }
  };
  return jsx(ThemeProvider, {
    theme: theme$1
  }, jsx("div", {
    sx: style$3.root
  }, jsx("button", {
    onClick: handleClick,
    disabled: isLoading,
    sx: _extends({}, style$3.button, buttonStyle)
  }, jsx("div", {
    sx: ringStyle
  }), jsx(ThumbIcon, {
    sx: iconStyle
  })), isCounterVisible && jsx("div", {
    sx: style$3.counter
  }, totalLikes)));
};

var TwitterHeartFull = function TwitterHeartFull(_ref) {
  var className = _ref.className;
  return React.createElement("svg", {
    viewBox: "0 0 24 24",
    className: className
  }, React.createElement("g", null, React.createElement("path", {
    d: "M12 21.638h-.014C9.403 21.59 1.95 14.856 1.95 8.478c0-3.064 2.525-5.754 5.403-5.754 2.29 0 3.83 1.58 4.646 2.73.814-1.148 2.354-2.73 4.645-2.73 2.88 0 5.404 2.69 5.404 5.755 0 6.376-7.454 13.11-10.037 13.157H12z"
  })));
};

var TwitterHeartEmpty = function TwitterHeartEmpty(_ref) {
  var className = _ref.className;
  return React.createElement("svg", {
    viewBox: "0 0 24 24",
    className: className
  }, React.createElement("g", null, React.createElement("path", {
    d: "M12 21.638h-.014C9.403 21.59 1.95 14.856 1.95 8.478c0-3.064 2.525-5.754 5.403-5.754 2.29 0 3.83 1.58 4.646 2.73.814-1.148 2.354-2.73 4.645-2.73 2.88 0 5.404 2.69 5.404 5.755 0 6.376-7.454 13.11-10.037 13.157H12zM7.354 4.225c-2.08 0-3.903 1.988-3.903 4.255 0 5.74 7.034 11.596 8.55 11.658 1.518-.062 8.55-5.917 8.55-11.658 0-2.267-1.823-4.255-3.903-4.255-2.528 0-3.94 2.936-3.952 2.965-.23.562-1.156.562-1.387 0-.014-.03-1.425-2.965-3.954-2.965z"
  })));
};

var style$4 = {
  root: {
    color: 'inherit',
    fontFamily: 'inherit',
    display: 'inline-flex',
    alignItems: 'center'
  },
  counter: {
    ml: '.4em',
    fontFamily: 'body',
    color: 'text'
  },
  icon: {
    height: '1.7em',
    display: 'block',
    margin: '0 auto',
    animationTimingFunction: 'cubic-bezier(0.280, 0.840, 0.420, 1)'
  },
  button: {
    border: 'none',
    outline: 'none',
    cursor: 'pointer',
    background: 'transparent',
    appearance: 'none',
    position: 'relative',
    fontSize: 'inherit',
    borderRadius: '100%',
    transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
    width: '3em',
    height: '3em',
    padding: '0',
    lineHeight: '1em',
    '@media (hover: hover)': {
      ':hover': {
        transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
        backgroundColor: '#E5214A20'
      }
    }
  },
  dotsRotContainer: {
    position: 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translate(-50%, -50%)'
  },
  dotsContainer: {
    position: 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translate(-50%, -50%) rotate(165deg)'
  },
  ring: {
    opacity: '0',
    width: "90%",
    height: "90%",
    borderRadius: "3em",
    boxSizing: 'border-box',
    position: 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translate(-50%, -50%)'
  }
};

var dotBurst = /*#__PURE__*/keyframes({
  '0%': {
    opacity: '0',
    transform: 'translateY(-0.5em) scale(1)'
  },
  '52%': {
    opacity: '1',
    transform: 'translateY(-1.6em) scale(1.3)'
  },
  '100%': {
    transform: 'translateY(-2.7em) scale(0)'
  }
});
var iconScale$4 = /*#__PURE__*/keyframes({
  '0%': {
    transform: 'scale(0)'
  },
  '52%': {
    transform: 'scale(1.2)'
  },
  '72%': {
    transform: 'scale(0.9)'
  },
  '100%': {
    transform: 'scale(1)'
  }
});
var ringScale$2 = /*#__PURE__*/keyframes({
  '0%': {
    opacity: '1',
    borderWidth: '0.5em'
  },
  '52%': {
    opacity: '0',
    borderWidth: '0'
  }
});
var DOTS = {
  count: 6,
  size: 0.2,
  colors: ['lightcoral', 'yellow', 'pink', 'lightseagreen', 'orange', 'aliceblue'],
  durationMs: 600,
  animation: dotBurst
};
var RING$2 = {
  color: '#e095ed',
  durationMs: DOTS.durationMs,
  animation: ringScale$2
};
var HEART = {
  animation: iconScale$4,
  durationMs: 700
};
var Twitter = function Twitter(_ref) {
  var isLoading = _ref.isLoading,
      userLiked = _ref.userLiked,
      totalLikes = _ref.totalLikes,
      handlePress = _ref.handlePress,
      isCounterVisible = _ref.isCounterVisible;

  var _useState = useState(false),
      animationActive = _useState[0],
      setAnimationActive = _useState[1];

  var dots = [];

  for (var i = 0; i < DOTS.count; i++) {
    dots.push(jsx("div", {
      key: i,
      sx: {
        transform: "rotate(" + 360 / DOTS.count * i + "deg)",
        transformOrigin: DOTS.size + "em 0",
        marginLeft: "-" + DOTS.size + "em"
      }
    }, jsx("div", {
      sx: {
        opacity: '0',
        width: DOTS.size + "em",
        height: DOTS.size + "em",
        borderRadius: DOTS.size + "em",
        backgroundColor: "" + DOTS.colors[i],
        position: 'absolute',
        animation: animationActive ? dotBurst + " " + DOTS.durationMs + "ms ease forwards" : null
      }
    })));
  }

  var ring = jsx("div", {
    sx: _extends({}, style$4.ring, {
      border: "1px solid " + RING$2.color,
      animation: animationActive ? RING$2.animation + " " + RING$2.durationMs + "ms ease forwards" : null
    })
  });
  var handleClick = useCallback(function (e) {
    if (!userLiked) {
      setAnimationActive(true);
      setTimeout(function () {
        return setAnimationActive(false);
      }, DOTS.durationMs);
    }

    handlePress(e);
  }, [userLiked, handlePress]);
  var Icon = userLiked ? TwitterHeartFull : TwitterHeartEmpty;
  return jsx("div", {
    sx: style$4.root
  }, jsx("button", {
    onClick: handleClick,
    sx: style$4.button,
    disabled: isLoading
  }, jsx("div", {
    sx: style$4.dotsContainer
  }, " ", dots, " "), jsx("div", {
    sx: style$4.dotsRotContainer
  }, " ", dots, " "), ring, jsx(Icon, {
    sx: _extends({}, style$4.icon, {
      fill: isLoading ? '#aaa' : '#E5214A',
      animation: animationActive ? HEART.animation + " " + HEART.durationMs + "ms ease forwards" : null
    })
  })), isCounterVisible && jsx("div", {
    sx: style$4.counter
  }, totalLikes));
};

var ChevronIcon = function ChevronIcon(_ref) {
  var className = _ref.className;
  return React.createElement("svg", {
    id: "Chevron",
    className: className,
    version: "1.1",
    xmlns: "http://www.w3.org/2000/svg",
    x: "0px",
    y: "0px",
    viewBox: "0 0 512 512",
    xmlSpace: "preserve"
  }, React.createElement("path", {
    d: "M505.183,123.179c-9.087-9.087-23.824-9.089-32.912,0.002l-216.266,216.27L39.729,123.179\n  \tc-9.087-9.087-23.824-9.089-32.912,0.002c-9.089,9.089-9.089,23.824,0,32.912L239.55,388.82c4.364,4.364,10.283,6.816,16.455,6.816\n  \tc6.172,0,12.092-2.453,16.455-6.817l232.721-232.727C514.272,147.004,514.272,132.268,505.183,123.179z"
  }));
};

var style$5 = {
  root: {
    color: 'inherit',
    fontFamily: 'inherit',
    display: 'inline-flex',
    alignItems: 'center'
  },
  counter: {
    color: 'text',
    fontFamily: 'body',
    mt: '.2em'
  },
  icon: {
    height: '0.8em',
    display: 'block',
    margin: '0 auto',
    fill: 'icon'
  },
  iconContainer: {
    transform: 'rotate(180deg)'
  },
  ring: {
    opacity: '0',
    width: "90%",
    height: "90%",
    borderRadius: "3em",
    border: "1px solid",
    borderColor: 'highlight',
    boxSizing: 'border-box',
    position: 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translate(-50%, -50%)'
  },
  button: {
    border: 'none',
    outline: 'none',
    boxShadow: '0 0 0',
    transform: 'translate3d(0,0,0)',
    cursor: 'pointer',
    appearance: 'none',
    position: 'relative',
    fontSize: 'inherit',
    borderRadius: '100%',
    transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
    width: '3em',
    height: '3em',
    padding: '0',
    lineHeight: '1em',
    bg: 'background',
    ':hover': {
      transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
      opacity: 0.75
    }
  }
};

var theme$2 = {
  buttons: {
    inactive: {
      backgroundColor: 'background'
    },
    active: {
      transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
      backgroundColor: 'primary'
    }
  }
};

var iconScale$5 = /*#__PURE__*/keyframes({
  '0%': {
    transform: 'scale(0)'
  },
  '52%': {
    transform: 'scale(1.2)'
  },
  '72%': {
    transform: 'scale(0.9)'
  },
  '100%': {
    transform: 'scale(1)'
  }
});
var ringScale$3 = /*#__PURE__*/keyframes({
  '0%': {
    opacity: '1',
    borderWidth: '0.5em'
  },
  '52%': {
    opacity: '0',
    borderWidth: '0'
  }
});
var RING$3 = {
  durationMs: 600,
  animation: ringScale$3
};
var ICON$4 = {
  animation: iconScale$5,
  durationMs: 700
};
var Chevron = function Chevron(_ref) {
  var isLoading = _ref.isLoading,
      userLiked = _ref.userLiked,
      totalLikes = _ref.totalLikes,
      handlePress = _ref.handlePress,
      isCounterVisible = _ref.isCounterVisible;

  var _useState = useState(false),
      animationActive = _useState[0],
      setAnimationActive = _useState[1];

  var handleClick = useCallback(function (e) {
    handlePress(e);

    if (!userLiked) {
      setAnimationActive(true);
      setTimeout(function () {
        return setAnimationActive(false);
      }, ICON$4.durationMs);
    }
  }, [userLiked, handlePress]);

  var iconStyle = _extends({}, style$5.icon, {
    animation: animationActive ? ICON$4.animation + " " + ICON$4.durationMs + "ms ease forwards" : null
  });

  var ringStyle = _extends({}, style$5.ring, {
    animation: animationActive ? RING$3.animation + " " + RING$3.durationMs + "ms ease forwards" : null
  });

  var buttonStyle = {
    variant: userLiked ? 'buttons.active' : 'buttons.inactive',
    '@media (hover: hover)': {
      ':hover': {
        bg: userLiked ? 'primary' : 'background'
      }
    }
  };
  return jsx(ThemeProvider, {
    theme: theme$2
  }, jsx("div", {
    sx: style$5.root
  }, jsx("button", {
    onClick: handleClick,
    disabled: isLoading,
    sx: _extends({}, style$5.button, buttonStyle)
  }, jsx("div", {
    sx: ringStyle
  }), jsx("div", {
    sx: style$5.iconContainer
  }, jsx(ChevronIcon, {
    sx: iconStyle
  })), isCounterVisible && jsx("div", {
    sx: style$5.counter
  }, totalLikes))));
};

var templates$1 = {
  Simple: Simple$1,
  Twitter: Twitter,
  Chevron: Chevron
};

var LikeButton = function LikeButton(_ref) {
  var id = _ref.id,
      namespace = _ref.namespace,
      hideCounterIfLessThan = _ref.hideCounterIfLessThan,
      children = _ref.children,
      component = _ref.component,
      onLoad = _ref.onLoad,
      onPress = _ref.onPress;
  var client = useContext(ClientContext);

  var _useState = useState(null),
      response = _useState[0],
      setResponse = _useState[1];

  useSafeEffect(function () {
    try {
      if (!client) {
        return Promise.resolve();
      }

      return Promise.resolve(_catch(function () {
        return Promise.resolve(client.likeButtons.info({
          id: id,
          namespace: namespace
        })).then(function (result) {
          setResponse(result.data);

          if (onLoad) {
            onLoad(camelizeKeys(result.data));
          }
        });
      }, function (error) {
        console.error('Lyket error:', error && error.errors[0].message);
        throw error;
      }));
    } catch (e) {
      return Promise.reject(e);
    }
  }, [client, id, namespace, onLoad, setResponse]);
  var handlePress = useCallback(function (e) {
    try {
      e.preventDefault();

      if (!client) {
        return Promise.resolve();
      }

      return Promise.resolve(_catch(function () {
        return Promise.resolve(client.likeButtons.press({
          id: id,
          namespace: namespace
        })).then(function (result) {
          setResponse(result.data);

          if (onPress) {
            onPress(camelizeKeys(result.data));
          }
        });
      }, function (error) {
        console.error('Lyket error:', error && error.errors[0].message);
        throw error;
      }));
    } catch (e) {
      return Promise.reject(e);
    }
  }, [client, id, namespace, onPress, setResponse]);
  var isCounterVisible = true;

  if (!response || hideCounterIfLessThan && response && response.attributes.total_likes < hideCounterIfLessThan) {
    isCounterVisible = false;
  }

  var props = {
    isLoading: !response,
    totalLikes: response && response.attributes.total_likes || 0,
    userLiked: response && response.attributes.user_has_liked || false,
    handlePress: handlePress,
    isCounterVisible: isCounterVisible
  };

  if (children) {
    return children(props);
  }

  var Component = component || Simple$1;
  return React.createElement(Component, Object.assign({}, props));
};

LikeButton.templates = templates$1;

var ThumbIcon$1 = function ThumbIcon(_ref) {
  var className = _ref.className;
  return React.createElement("svg", {
    viewBox: "0 0 215.34 215.34",
    className: className
  }, React.createElement("path", {
    d: "M7.67,93.896h52.077V215.34H7.67V93.896z M207.67,106.682c-7.189-7.189-14.382-14.379-21.572-21.57h-41.451l9.877-25.102  l2.801-7.119l-0.258-2.049L151,2.33L121.115,0L110.15,37.822L80.954,85.176V215.34h107.89L207.67,106.682z"
  }));
};

var style$6 = {
  root: {
    color: 'inherit',
    fontFamily: 'inherit',
    display: 'inline-flex',
    alignItems: 'center'
  },
  counter: {
    ml: '.5em',
    mr: '.5em',
    minWidth: '2em',
    color: 'text',
    textAlign: 'center',
    fontFamily: 'body',
    fontWeight: 'body'
  },
  icon: {
    height: '1em',
    display: 'block',
    margin: '0 auto',
    fill: 'icon'
  },
  ring: {
    opacity: '0',
    width: "90%",
    height: "90%",
    borderRadius: "3em",
    border: "1px solid",
    borderColor: 'highlight',
    boxSizing: 'border-box',
    position: 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translate(-50%, -50%)'
  },
  button: {
    border: 'none',
    outline: 'none',
    cursor: 'pointer',
    appearance: 'none',
    position: 'relative',
    fontSize: 'inherit',
    borderRadius: '100%',
    width: '3em',
    height: '3em',
    padding: '0',
    lineHeight: '1em',
    bg: 'background',
    transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
    opacity: 1,
    boxShadow: '0 0 0',
    transform: 'translate3d(0,0,0)',
    ':hover': {
      transition: 'all 250ms cubic-bezier(0.4, 0.0, 0.2, 1) 0ms',
      opacity: 0.75
    }
  }
};

var theme$3 = {
  buttons: {
    upInactive: {
      backgroundColor: 'background'
    },
    upActive: {
      backgroundColor: 'primary',
      transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms'
    },
    downInactive: {
      backgroundColor: 'background'
    },
    downActive: {
      transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
      backgroundColor: 'secondary'
    }
  }
};

var iconScale$6 = /*#__PURE__*/keyframes({
  '0%': {
    transform: 'scale(0)'
  },
  '52%': {
    transform: 'scale(1.2)'
  },
  '72%': {
    transform: 'scale(0.9)'
  },
  '100%': {
    transform: 'scale(1)'
  }
});
var ringScale$4 = /*#__PURE__*/keyframes({
  '0%': {
    opacity: '1',
    borderWidth: '0.5em'
  },
  '52%': {
    opacity: '0',
    borderWidth: '0'
  }
});
var RING$4 = {
  durationMs: 600,
  animation: ringScale$4
};
var ICON$5 = {
  animation: iconScale$6,
  durationMs: 700
};
var Simple$2 = function Simple(_ref) {
  var isLoading = _ref.isLoading,
      userVoteDirection = _ref.userVoteDirection,
      totalScore = _ref.totalScore,
      handlePressUp = _ref.handlePressUp,
      handlePressDown = _ref.handlePressDown,
      isCounterVisible = _ref.isCounterVisible;

  var _useState = useState(false),
      animationActiveUp = _useState[0],
      setAnimationActiveUp = _useState[1];

  var _useState2 = useState(false),
      animationActiveDown = _useState2[0],
      setAnimationActiveDown = _useState2[1];

  var handleClickUp = useCallback(function (e) {
    handlePressUp(e);

    if (userVoteDirection !== 1) {
      setAnimationActiveUp(true);
      setTimeout(function () {
        return setAnimationActiveUp(false);
      }, ICON$5.durationMs);
    }
  }, [userVoteDirection, handlePressUp]);
  var handleClickDown = useCallback(function (e) {
    handlePressDown(e);

    if (userVoteDirection !== -1) {
      setAnimationActiveDown(true);
      setTimeout(function () {
        return setAnimationActiveDown(false);
      }, ICON$5.durationMs);
    }
  }, [userVoteDirection, handlePressDown]);

  var iconStyleUp = _extends({}, style$6.icon, {
    animation: animationActiveUp ? ICON$5.animation + " " + ICON$5.durationMs + "ms ease forwards" : null
  });

  var iconStyleDown = _extends({}, style$6.icon, {
    animation: animationActiveDown ? ICON$5.animation + " " + ICON$5.durationMs + "ms ease forwards" : null
  });

  var ringStyleUp = _extends({}, style$6.ring, {
    animation: animationActiveUp ? RING$4.animation + " " + RING$4.durationMs + "ms ease forwards" : null
  });

  var ringStyleDown = _extends({}, style$6.ring, {
    animation: animationActiveDown ? RING$4.animation + " " + RING$4.durationMs + "ms ease forwards" : null
  });

  var upButtonStyle = {
    variant: userVoteDirection === 1 ? 'buttons.upActive' : 'buttons.upInactive',
    '@media (hover: hover)': {
      ':hover': {
        bg: userVoteDirection === 1 ? 'primary' : 'background'
      }
    }
  };
  var downButtonStyle = {
    variant: userVoteDirection === -1 ? 'buttons.downActive' : 'buttons.downInactive',
    '@media (hover: hover)': {
      ':hover': {
        bg: userVoteDirection === -1 ? 'secondary' : 'background'
      }
    }
  };
  return jsx(ThemeProvider, {
    theme: theme$3
  }, jsx("div", {
    sx: style$6.root
  }, jsx("button", {
    onClick: handleClickUp,
    disabled: isLoading,
    sx: _extends({}, style$6.button, upButtonStyle)
  }, jsx("div", {
    sx: ringStyleUp
  }), jsx(ThumbIcon$1, {
    sx: iconStyleUp
  })), isCounterVisible && jsx("div", {
    sx: style$6.counter
  }, totalScore), isLoading && jsx("div", {
    sx: style$6.counter
  }, "-"), jsx("button", {
    onClick: handleClickDown,
    disabled: isLoading,
    sx: _extends({}, style$6.button, downButtonStyle)
  }, jsx("div", {
    sx: ringStyleDown
  }), jsx("div", {
    style: {
      transform: 'rotate(180deg)'
    }
  }, jsx(ThumbIcon$1, {
    sx: iconStyleDown
  })))));
};

var ChevronIcon$1 = function ChevronIcon(_ref) {
  var className = _ref.className;
  return React.createElement("svg", {
    id: "Chevron",
    className: className,
    version: "1.1",
    xmlns: "http://www.w3.org/2000/svg",
    x: "0px",
    y: "0px",
    viewBox: "0 0 512 512",
    xmlSpace: "preserve"
  }, React.createElement("path", {
    d: "M505.183,123.179c-9.087-9.087-23.824-9.089-32.912,0.002l-216.266,216.27L39.729,123.179\n  \tc-9.087-9.087-23.824-9.089-32.912,0.002c-9.089,9.089-9.089,23.824,0,32.912L239.55,388.82c4.364,4.364,10.283,6.816,16.455,6.816\n  \tc6.172,0,12.092-2.453,16.455-6.817l232.721-232.727C514.272,147.004,514.272,132.268,505.183,123.179z"
  }));
};

var style$7 = {
  root: {
    color: 'inherit',
    fontFamily: 'inherit',
    display: 'inline-flex',
    alignItems: 'center'
  },
  counter: {
    ml: '.5em',
    mr: '.5em',
    minWidth: '2em',
    color: 'text',
    textAlign: 'center',
    fontFamily: 'body'
  },
  icon: {
    height: '1em',
    display: 'block',
    margin: '0 auto',
    fill: 'icon'
  },
  ring: {
    opacity: '0',
    width: "90%",
    height: "90%",
    borderRadius: "3em",
    border: "1px solid",
    borderColor: 'highlight',
    boxSizing: 'border-box',
    position: 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translate(-50%, -50%)'
  },
  button: {
    border: 'none',
    outline: 'none',
    cursor: 'pointer',
    appearance: 'none',
    position: 'relative',
    fontSize: 'inherit',
    borderRadius: '100%',
    width: '3em',
    height: '3em',
    padding: '0',
    lineHeight: '1em',
    bg: 'background',
    transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
    opacity: 1,
    boxShadow: '0 0 0',
    transform: 'translate3d(0,0,0)',
    ':hover': {
      transition: 'all 250ms cubic-bezier(0.4, 0.0, 0.2, 1) 0ms',
      opacity: 0.75
    }
  }
};

var theme$4 = {
  buttons: {
    upInactive: {
      backgroundColor: 'background'
    },
    upActive: {
      backgroundColor: 'primary',
      transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms'
    },
    downInactive: {
      backgroundColor: 'background'
    },
    downActive: {
      transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
      backgroundColor: 'secondary'
    }
  }
};

var iconScale$7 = /*#__PURE__*/keyframes({
  '0%': {
    transform: 'scale(0)'
  },
  '52%': {
    transform: 'scale(1.2)'
  },
  '72%': {
    transform: 'scale(0.9)'
  },
  '100%': {
    transform: 'scale(1)'
  }
});
var ringScale$5 = /*#__PURE__*/keyframes({
  '0%': {
    opacity: '1',
    borderWidth: '0.5em'
  },
  '52%': {
    opacity: '0',
    borderWidth: '0'
  }
});
var RING$5 = {
  durationMs: 600,
  animation: ringScale$5
};
var ICON$6 = {
  animation: iconScale$7,
  durationMs: 700
};
var Chevron$1 = function Chevron(_ref) {
  var isLoading = _ref.isLoading,
      userVoteDirection = _ref.userVoteDirection,
      totalScore = _ref.totalScore,
      handlePressUp = _ref.handlePressUp,
      handlePressDown = _ref.handlePressDown,
      isCounterVisible = _ref.isCounterVisible;

  var _useState = useState(false),
      animationActiveUp = _useState[0],
      setAnimationActiveUp = _useState[1];

  var _useState2 = useState(false),
      animationActiveDown = _useState2[0],
      setAnimationActiveDown = _useState2[1];

  var handleClickUp = useCallback(function (e) {
    handlePressUp(e);

    if (userVoteDirection !== 1) {
      setAnimationActiveUp(true);
      setTimeout(function () {
        return setAnimationActiveUp(false);
      }, ICON$6.durationMs);
    }
  }, [userVoteDirection, handlePressUp]);
  var handleClickDown = useCallback(function (e) {
    handlePressDown(e);

    if (userVoteDirection !== -1) {
      setAnimationActiveDown(true);
      setTimeout(function () {
        return setAnimationActiveDown(false);
      }, ICON$6.durationMs);
    }
  }, [userVoteDirection, handlePressDown]);

  var iconStyleUp = _extends({}, style$7.icon, {
    animation: animationActiveUp ? ICON$6.animation + " " + ICON$6.durationMs + "ms ease forwards" : null
  });

  var iconStyleDown = _extends({}, style$7.icon, {
    animation: animationActiveDown ? ICON$6.animation + " " + ICON$6.durationMs + "ms ease forwards" : null
  });

  var ringStyleUp = _extends({}, style$7.ring, {
    animation: animationActiveUp ? RING$5.animation + " " + RING$5.durationMs + "ms ease forwards" : null
  });

  var ringStyleDown = _extends({}, style$7.ring, {
    animation: animationActiveDown ? RING$5.animation + " " + RING$5.durationMs + "ms ease forwards" : null
  });

  var upButtonStyle = {
    variant: userVoteDirection === 1 ? 'buttons.upActive' : 'buttons.upInactive',
    '@media (hover: hover)': {
      ':hover': {
        bg: userVoteDirection === 1 ? 'primary' : 'background'
      }
    }
  };
  var downButtonStyle = {
    variant: userVoteDirection === -1 ? 'buttons.downActive' : 'buttons.downInactive',
    '@media (hover: hover)': {
      ':hover': {
        bg: userVoteDirection === -1 ? 'secondary' : 'background'
      }
    }
  };
  return jsx(ThemeProvider, {
    theme: theme$4
  }, jsx("div", {
    sx: style$7.root
  }, jsx("button", {
    onClick: handleClickUp,
    disabled: isLoading,
    sx: _extends({}, style$7.button, upButtonStyle)
  }, jsx("div", {
    sx: ringStyleUp
  }), jsx("div", {
    style: {
      transform: 'rotate(180deg)'
    }
  }, jsx(ChevronIcon$1, {
    sx: iconStyleUp
  }))), isCounterVisible && jsx("div", {
    sx: style$7.counter
  }, totalScore), isLoading && jsx("div", {
    sx: style$7.counter
  }, "-"), jsx("button", {
    onClick: handleClickDown,
    disabled: isLoading,
    sx: _extends({}, style$7.button, downButtonStyle)
  }, jsx("div", {
    sx: ringStyleDown
  }), jsx(ChevronIcon$1, {
    sx: iconStyleDown
  }))));
};

var RedditArrow = function RedditArrow(_ref) {
  var className = _ref.className;
  return React.createElement("svg", {
    version: "1.0",
    viewBox: "0 0 1280 1280",
    className: className,
    preserveAspectRatio: "xMidYMid meet"
  }, React.createElement("g", {
    transform: "translate(0,1280) scale(0.100000,-0.100000)",
    stroke: "none"
  }, React.createElement("path", {
    d: "M6245 12778 c-200 -44 -333 -117 -494 -272 -53 -51 -1149 -1102\n  -2436 -2336 -3028 -2904 -3071 -2945 -3122 -3010 -133 -170 -219 -465 -183\n  -629 35 -159 45 -192 80 -267 50 -109 104 -185 187 -265 83 -81 140 -119 246\n  -169 175 -81 33 -74 1677 -77 1359 -3 1466 -4 1478 -20 9 -13 12 -527 12\n  -2487 0 -2658 -2 -2524 50 -2663 84 -226 259 -410 479 -503 40 -17 123 -42\n  184 -56 l112 -24 1905 3 c1897 2 1905 2 1995 23 173 41 319 120 439 238 104\n  102 171 212 222 366 l29 85 5 2507 c3 1701 8 2510 15 2517 7 7 487 12 1475 13\n  1644 4 1502 -3 1677 78 104 48 161 88 247 169 137 131 213 277 256 491 25 119\n  25 151 0 270 -32 159 -75 262 -155 375 -53 75 -26 49 -3010 2910 -1353 1297\n  -2506 2402 -2561 2456 -169 163 -299 234 -509 278 -119 25 -181 25 -300 -1z"
  })));
};

var style$8 = {
  root: {
    color: 'inherit',
    fontFamily: 'inherit',
    display: 'inline-flex',
    alignItems: 'center',
    flexDirection: 'column'
  },
  counter: {
    mt: '.2em',
    mb: '.2em',
    minWidth: '2em',
    textAlign: 'center',
    fontFamily: 'body'
  },
  icon: {
    height: '1em',
    display: 'block',
    margin: '0 auto',
    animationTimingFunction: 'cubic-bezier(0.280, 0.840, 0.420, 1)'
  },
  button: {
    border: 'none',
    outline: 'none',
    cursor: 'pointer',
    background: 'transparent',
    appearance: 'none',
    position: 'relative',
    fontSize: 'inherit',
    borderRadius: '3px',
    transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
    width: '2em',
    height: '2em',
    padding: '0',
    lineHeight: '1em'
  }
};

var theme$5 = {
  colors: {
    lightGray: '#25252510',
    darkGray: '#878a8c',
    blue: '#5a75cc',
    red: '#cc3700'
  },
  buttons: {
    downInactive: {
      svg: {
        fill: 'darkGray'
      }
    },
    upInactive: {
      svg: {
        fill: 'darkGray'
      }
    },
    upActive: {
      backgroundColor: 'lightGray',
      svg: {
        fill: 'red'
      }
    },
    downActive: {
      backgroundColor: 'lightGray',
      svg: {
        fill: 'blue'
      }
    }
  }
};

var iconScale$8 = /*#__PURE__*/keyframes({
  '0%': {
    transform: 'scale(0)'
  },
  '52%': {
    transform: 'scale(1.2)'
  },
  '72%': {
    transform: 'scale(0.9)'
  },
  '100%': {
    transform: 'scale(1)'
  }
});
var ICON$7 = {
  animation: iconScale$8,
  durationMs: 700
};
var Reddit = function Reddit(_ref) {
  var isLoading = _ref.isLoading,
      userVoteDirection = _ref.userVoteDirection,
      totalScore = _ref.totalScore,
      handlePressUp = _ref.handlePressUp,
      handlePressDown = _ref.handlePressDown,
      isCounterVisible = _ref.isCounterVisible;

  var _useState = useState(false),
      animationActiveUp = _useState[0],
      setAnimationActiveUp = _useState[1];

  var _useState2 = useState(false),
      animationActiveDown = _useState2[0],
      setAnimationActiveDown = _useState2[1];

  var handleClickUp = useCallback(function (e) {
    handlePressUp(e);

    if (userVoteDirection !== 1) {
      setAnimationActiveUp(true);
      setTimeout(function () {
        return setAnimationActiveUp(false);
      }, ICON$7.durationMs);
    }
  }, [userVoteDirection, handlePressUp]);
  var handleClickDown = useCallback(function (e) {
    handlePressDown(e);

    if (userVoteDirection !== -1) {
      setAnimationActiveDown(true);
      setTimeout(function () {
        return setAnimationActiveDown(false);
      }, ICON$7.durationMs);
    }
  }, [userVoteDirection, handlePressDown]);

  var iconStyleUp = _extends({}, style$8.icon, {
    animation: animationActiveUp ? ICON$7.animation + " " + ICON$7.durationMs + "ms ease forwards" : null
  });

  var iconStyleDown = _extends({}, style$8.icon, {
    animation: animationActiveDown ? ICON$7.animation + " " + ICON$7.durationMs + "ms ease forwards" : null
  });

  return jsx(ThemeProvider, {
    theme: theme$5
  }, jsx("div", {
    sx: style$8.root
  }, jsx("button", {
    onClick: handleClickUp,
    disabled: isLoading,
    sx: _extends({}, style$8.button, {
      variant: userVoteDirection === 1 ? 'buttons.upActive' : 'buttons.upInactive',
      '@media (hover: hover)': {
        ':hover': {
          variant: 'buttons.upActive',
          transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms'
        }
      }
    })
  }, jsx(RedditArrow, {
    sx: iconStyleUp
  })), isCounterVisible && jsx("div", {
    sx: style$8.counter
  }, totalScore), isLoading && jsx("div", {
    sx: style$8.counter
  }, "-"), jsx("button", {
    onClick: handleClickDown,
    disabled: isLoading,
    sx: _extends({}, style$8.button, {
      transform: 'rotate(180deg)',
      variant: userVoteDirection === -1 ? 'buttons.downActive' : 'buttons.downInactive',
      '@media (hover: hover)': {
        ':hover': {
          variant: 'buttons.downActive',
          transition: 'all 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms'
        }
      }
    })
  }, jsx(RedditArrow, {
    sx: iconStyleDown
  }))));
};

var templates$2 = {
  Simple: Simple$2,
  Chevron: Chevron$1,
  Reddit: Reddit
};

var UpdownButton = function UpdownButton(_ref) {
  var id = _ref.id,
      namespace = _ref.namespace,
      hideCounterIfLessThan = _ref.hideCounterIfLessThan,
      children = _ref.children,
      component = _ref.component,
      onLoad = _ref.onLoad,
      onPressUp = _ref.onPressUp,
      onPressDown = _ref.onPressDown;
  var client = useContext(ClientContext);

  var _useState = useState(null),
      response = _useState[0],
      setResponse = _useState[1];

  useSafeEffect(function () {
    try {
      return Promise.resolve(_catch(function () {
        var _temp = function () {
          if (client) {
            return Promise.resolve(client.updownButtons.info({
              id: id,
              namespace: namespace
            })).then(function (result) {
              setResponse(result.data);

              if (onLoad) {
                onLoad(camelizeKeys(result.data));
              }
            });
          }
        }();

        if (_temp && _temp.then) return _temp.then(function () {});
      }, function (error) {
        console.error('Lyket error:', error && error.errors[0].message);
        throw error;
      }));
    } catch (e) {
      return Promise.reject(e);
    }
  }, [client, id, namespace, onLoad]);
  var handlePressUp = useCallback(function (e) {
    try {
      e.preventDefault();
      return Promise.resolve(_catch(function () {
        var _temp2 = function () {
          if (client) {
            return Promise.resolve(client.updownButtons.pressUp({
              id: id,
              namespace: namespace
            })).then(function (result) {
              setResponse(result.data);

              if (onPressUp) {
                onPressUp(camelizeKeys(result.data));
              }
            });
          }
        }();

        if (_temp2 && _temp2.then) return _temp2.then(function () {});
      }, function (error) {
        console.error('Lyket error:', error && error.errors[0].message);
        throw error;
      }));
    } catch (e) {
      return Promise.reject(e);
    }
  }, [client, id, namespace, onPressUp]);
  var handlePressDown = useCallback(function (e) {
    try {
      e.preventDefault();
      return Promise.resolve(_catch(function () {
        var _temp3 = function () {
          if (client) {
            return Promise.resolve(client.updownButtons.pressDown({
              id: id,
              namespace: namespace
            })).then(function (result) {
              setResponse(result.data);

              if (onPressDown) {
                onPressDown(camelizeKeys(result.data));
              }
            });
          }
        }();

        if (_temp3 && _temp3.then) return _temp3.then(function () {});
      }, function (error) {
        console.error('Lyket error:', error && error.errors[0].message);
        throw error;
      }));
    } catch (e) {
      return Promise.reject(e);
    }
  }, [client, id, namespace, onPressDown]);
  var isCounterVisible = true;

  if (!response || hideCounterIfLessThan && response && response.attributes.total_score < hideCounterIfLessThan) {
    isCounterVisible = false;
  }

  var props = {
    isLoading: !response,
    totalScore: response && response.attributes.total_score || 0,
    userVoteDirection: response && response.attributes.user_vote_direction || 0,
    handlePressUp: handlePressUp,
    handlePressDown: handlePressDown,
    isCounterVisible: isCounterVisible
  };

  if (children) {
    return children(props);
  }

  var Component = component || Simple$2;
  return React.createElement(Component, Object.assign({}, props));
};

UpdownButton.templates = templates$2;

export { ClapButton, Client, LikeButton, Provider, UpdownButton };
