// @theme-ui/mdx/dist/index.umd.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@theme-ui/core'), require('@theme-ui/css'), require('@emotion/styled'), require('@mdx-js/react')) :
  typeof define === 'function' && define.amd ? define(['exports', '@theme-ui/core', '@theme-ui/css', '@emotion/styled', '@mdx-js/react'], factory) :
  (global = global || self, factory(global.mdx = {}, global.core, global.css, global.styled, global.react));
}(this, (function (exports, core, css, styled, react) {
  function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

  var styled__default = /*#__PURE__*/_interopDefaultLegacy(styled);

  function _extends() {
    _extends = Object.assign || function (target) {
      for (var i = 1; i < arguments.length; i++) {
        var source = arguments[i];

        for (var key in source) {
          if (Object.prototype.hasOwnProperty.call(source, key)) {
            target[key] = source[key];
          }
        }
      }

      return target;
    };

    return _extends.apply(this, arguments);
  }

  function _objectWithoutPropertiesLoose(source, excluded) {
    if (source == null) return {};
    var target = {};
    var sourceKeys = Object.keys(source);
    var key, i;

    for (i = 0; i < sourceKeys.length; i++) {
      key = sourceKeys[i];
      if (excluded.indexOf(key) >= 0) continue;
      target[key] = source[key];
    }

    return target;
  }

  var tags = ['p', 'b', 'i', 'a', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'img', 'pre', 'code', 'ol', 'ul', 'li', 'blockquote', 'hr', 'em', 'table', 'tr', 'th', 'td', 'em', 'strong', 'del', // mdx
  'inlineCode', 'thematicBreak', // other
  'div', // theme-ui
  'root'];
  var aliases = {
    inlineCode: 'code',
    thematicBreak: 'hr',
    root: 'div'
  };

  var isAlias = function isAlias(x) {
    return x in aliases;
  };

  var alias = function alias(n) {
    return isAlias(n) ? aliases[n] : n;
  };

  var propOverrides = {
    th: {
      align: 'textAlign'
    },
    td: {
      align: 'textAlign'
    }
  };
  var themed = function themed(key) {
    return function (_ref) {
      var theme = _ref.theme,
          rest = _objectWithoutPropertiesLoose(_ref, ["theme"]);

      var propsStyle = propOverrides[key];
      var extraStyles = propsStyle ? Object.keys(rest).filter(function (prop) {
        return propsStyle[prop] !== undefined;
      }).reduce(function (acc, prop) {
        var _extends2;

        return _extends({}, acc, (_extends2 = {}, _extends2[propsStyle[prop]] = rest[prop], _extends2));
      }, {}) : undefined;
      return css.css(_extends({}, css.get(theme, "styles." + key), extraStyles))(theme);
    };
  };
  var Themed = styled__default['default']('div')(themed('div'));
  var components = {};
  tags.forEach(function (tag) {
    // fixme?
    components[tag] = styled__default['default'](alias(tag))(themed(tag));
    Themed[tag] = components[tag];
  });

  var createComponents = function createComponents(comps) {
    var next = _extends({}, components);

    var componentKeys = Object.keys(comps);
    componentKeys.forEach(function (key) {
      next[key] = styled__default['default'](comps[key])(themed(key));
    });
    return next;
  };

  var MDXProvider = function MDXProvider(_ref2) {
    var components = _ref2.components,
        children = _ref2.children;
    var outer = react.useMDXComponents();
    return core.jsx(react.MDXProvider, {
      components: createComponents(_extends({}, outer, components)),
      children: children
    });
  };

  exports.MDXProvider = MDXProvider;
  exports.Themed = Themed;
  exports.components = components;
  exports.themed = themed;

})));
